diff -Naur --exclude .svn bcm5700-8.2.18/b57um.c bcm5700-8.2.18-ps/b57um.c
--- bcm5700-8.2.18/b57um.c	2005-08-02 02:30:49.000000000 +0200
+++ bcm5700-8.2.18-ps/b57um.c	2005-10-13 16:59:38.002818762 +0200
@@ -12,7 +12,7 @@
 
 
 char bcm5700_driver[] = "bcm5700";
-char bcm5700_version[] = "8.2.18";
+char bcm5700_version[] = "8.2.18_ps4";
 char bcm5700_date[] = "(08/01/05)";
 
 #define B57UM
@@ -64,7 +64,8 @@
 	{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
 #else
 static unsigned int adaptive_coalesce[MAX_UNITS] =
-	{1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};
+/*	{1,1,1,1,1,1,1,1,1,1,1,1,1,1,1,1};*/
+	{0,0,0,0,0,0,0,0,0,0,0,0,0,0,0,0};
 #endif
 
 #define RX_COAL_TK DEFAULT_RX_COALESCING_TICKS
@@ -279,6 +280,66 @@
 #define spin_is_locked(lock)    (test_bit(0,(lock)))
 #endif
 
+/* ParaStation extension start */
+
+#ifdef BCM_NAPI_RXPOLL
+#warning "BCM_NAPI_RXPOLL might not work with ParaStation!!!"
+#endif
+#define MAX_bcm5700_irq_poll 16
+
+static struct {
+	struct net_device *netdev;
+} bcm5700_irq_poll[MAX_bcm5700_irq_poll];
+
+static int bcm5700_irq_poll_last = 0;
+
+static void
+bcm5700_add_poll(struct net_device *netdev)
+{
+	int i;
+	
+	if (bcm5700_irq_poll_last == MAX_bcm5700_irq_poll - 1)
+		return;
+
+	for (i = 0; i < bcm5700_irq_poll_last; i++) {
+		if (bcm5700_irq_poll[i].netdev == netdev)
+			break;
+	}
+	bcm5700_irq_poll[i].netdev = netdev;
+
+	if (i == bcm5700_irq_poll_last) 
+		bcm5700_irq_poll_last++;
+}
+
+static void
+bcm5700_del_poll(struct net_device *netdev)
+{
+	int i;
+	for (i = 0; i < bcm5700_irq_poll_last; i++) {
+		if (bcm5700_irq_poll[i].netdev == netdev)
+			break;
+	}
+
+	if ((i < bcm5700_irq_poll_last) &&
+	    (bcm5700_irq_poll[i].netdev == netdev)) {
+		bcm5700_irq_poll_last--;
+		bcm5700_irq_poll[i].netdev =
+			bcm5700_irq_poll[bcm5700_irq_poll_last].netdev;
+	}
+}
+
+#include "p4ether_pub.h"
+static char vcid_ps4poll[] __attribute__(( unused )) =
+"$Id: b57um.c 3604 2005-04-26 17:32:58Z hauke $";
+
+#ifdef BCM_NAPI_RXPOLL
+p4ether_netif_rx_t *bcm5700_netif_rx = netif_receive_skb;
+#else
+p4ether_netif_rx_t *bcm5700_netif_rx = netif_rx;
+#endif
+
+/* ParaStation extension end */
+
 #define BCM5700_LOCK(pUmDevice, flags)					\
 	if ((pUmDevice)->do_global_lock) {				\
 		spin_lock_irqsave(&(pUmDevice)->global_lock, flags);	\
@@ -1524,6 +1585,7 @@
 
 	netif_start_queue(dev);
 
+	bcm5700_add_poll(dev); /* ParaStation add polling. */
 
 	return 0;
 }
@@ -2187,8 +2249,9 @@
 	}
 
 	if (test_and_set_bit(0, (void*)&pUmDevice->interrupt)) {
-		printk(KERN_ERR "%s: Duplicate entry of the interrupt handler\n",
-			dev->name);
+		/* ParaStation need reentrant bcm5700_interrupt()!
+		   printk(KERN_ERR "%s: Duplicate entry of the interrupt handler\n",
+		   dev->name); */
 		bcm5700_intr_unlock(pUmDevice);
 		handled = 0;
 		return IRQ_RETVAL(handled);
@@ -2341,6 +2404,8 @@
 #if (LINUX_VERSION_CODE < 0x02032b)
 	dev->start = 0;
 #endif
+	bcm5700_del_poll(dev); /* ParaStation del polling */
+
 	netif_stop_queue(dev);
 	pUmDevice->opened = 0;
 
@@ -4423,6 +4488,32 @@
 #endif
 }
 
+/* ParaStation extension start */
+
+void
+bcm5700_polling(void)
+{
+	int i;
+
+	for (i = 0; i < bcm5700_irq_poll_last; i++) {
+		struct net_device *netdev = bcm5700_irq_poll[i].netdev;
+
+		/* bcm5700_interrupt dont use irq or regs and
+		   are surrouned by a lock
+		   (test_and_set_bit(0, (void*)&pUmDevice->interrupt)). */
+		bcm5700_interrupt(0, netdev, NULL);
+	}
+}
+
+#if LINUX_VERSION_CODE < KERNEL_VERSION(2,6,0)
+EXPORT_SYMBOL_NOVERS(bcm5700_polling);
+EXPORT_SYMBOL_NOVERS(bcm5700_netif_rx);
+#else
+EXPORT_SYMBOL(bcm5700_polling);
+EXPORT_SYMBOL(bcm5700_netif_rx);
+#endif
+
+/* ParaStation extension end */
 
 static struct pci_driver bcm5700_pci_driver = {
 	name:		bcm5700_driver,
@@ -5119,11 +5210,15 @@
 			else
 #endif
 			{
+#if 0
 #ifdef BCM_NAPI_RXPOLL
 				netif_receive_skb(skb);
 #else
 				netif_rx(skb);
 #endif
+#endif
+				bcm5700_netif_rx(skb); /* ParaStation Hook for fast-receive */
+
 			}
 		}
 		pUmDevice->dev->last_rx = jiffies;
diff -Naur --exclude .svn bcm5700-8.2.18/Makefile bcm5700-8.2.18-ps/Makefile
--- bcm5700-8.2.18/Makefile	2005-08-02 02:30:49.000000000 +0200
+++ bcm5700-8.2.18-ps/Makefile	2005-10-13 16:33:18.018821646 +0200
@@ -1,165 +1,11 @@
-#!/usr/bin/make
-# Makefile for building Linux Broadcom Gigabit ethernet driver as a module.
-# $id$
-
-# PREFIX may be set by the RPM build to set the effective root.
-PREFIX=
-ifeq ($(shell ls /lib/modules/$(shell uname -r)/build > /dev/null 2>&1 && echo build),)
-  ifeq ($(shell ls /usr/src/linux > /dev/null 2>&1 && echo linux),)
-    LINUX=
-  else
-    LINUX=/usr/src/linux
-  endif
-else
-  LINUX=/lib/modules/$(shell uname -r)/build
-endif
-
-ifneq ($(shell grep netdump_mode $(LINUX)/include/linux/kernel.h > /dev/null 2>&1 && echo rh),)
-  NETDUMP_CFLAG = -DRED_HAT_LINUX_KERNEL
-else
-  NETDUMP_CFLAG =
-endif
-
-# check if 2.4 kernel or 2.5+ kernel
-BCM_KVER:=$(shell uname -r | cut -c1-3 | sed 's/2\.[56]/2\.6/')
-
-ifeq ($(BCM_KVER), 2.6)
-# Makefile for 2.5+ kernel
-
-BCM_DRV = bcm5700.ko
-
-ifneq ($(KERNELRELEASE),)
+# $Id: Makefile 3431 2004-10-26 13:12:48Z hauke $
 
 obj-m += bcm5700.o
 
 bcm5700-objs    := b57um.o b57proc.o tigon3.o autoneg.o 5701rls.o tcp_seg.o b57diag.o
 
-EXTRA_CFLAGS = -DDBG=0 -DT3_JUMBO_RCV_RCB_ENTRY_COUNT=256 -DNICE_SUPPORT -DPCIX_TARGET_WORKAROUND=1 -DINCLUDE_TBI_SUPPORT -DINCLUDE_5701_AX_FIX=1 $(NETDUMP_CFLAG)
-
-else
-
-default:
-	make -C $(LINUX) SUBDIRS=$(shell pwd) modules
-
-endif
-
-else # ifeq ($(BCM_KVER),2.6)
-# Makefile for 2.4 kernel
-
-BCM_DRV = bcm5700.o
-
-ifeq ($(LINUX),)
-  $(error Linux kernel source tree not found)
-endif
-
-ifeq ($(shell which kgcc > /dev/null 2>&1 && echo kgcc),)
-  CC = gcc
-else
-  ifneq ($(findstring gcc version egcs-2.91.66, $(shell cat /proc/version 2>&1)),)
-    CC = kgcc
-  else
-    CC = gcc
-  endif
-endif
-
-LD = ld
-CFLAGS=-DMODULE -D__KERNEL__ -DDBG=0 -DT3_JUMBO_RCV_RCB_ENTRY_COUNT=256 -DNICE_SUPPORT -DPCIX_TARGET_WORKAROUND=1 -DINCLUDE_TBI_SUPPORT -DINCLUDE_5701_AX_FIX=1 $(NETDUMP_CFLAG) -Wall -Wstrict-prototypes -I$(LINUX)/include
-
-ifeq ($(wildcard ~/rpmdir),)
-  rpmdir = /usr/src/redhat
-else
-  rpmdir = $(wildcard ~/rpmdir)
-endif
-
-ARCH:=$(shell uname -m)
-
-ifeq ($(ARCH),x86_64)
-  CFLAGS+=-mno-red-zone -mcmodel=kernel -pipe -finline-limit=2000
-endif
-
-ifeq ($(ARCH),ia64)
-  CFLAGS+=-pipe -ffixed-r13 -mfixed-range=f10-f15,f32-f127 -falign-functions=32
-endif
-
-ifeq ($(ARCH),ppc64)
-  ifneq ($(shell ls /opt/cross/bin/powerpc64-linux-gcc > /dev/null 2>&1 && echo ppcgcc),)
-    CC=/opt/cross/bin/powerpc64-linux-gcc
-  else
-    CFLAGS+=-m64 
-  endif
-
-  ifneq ($(shell ls /opt/cross/bin/powerpc64-linux-ld > /dev/null 2>&1 && echo ppcld),)
-    LD=/opt/cross/bin/powerpc64-linux-ld
-  else
-    LD=ld -m elf64ppc --oformat elf64-powerpc -r -o
-  endif
-
-  CFLAGS+=-fno-strict-aliasing -fno-common -fomit-frame-pointer -msoft-float -pipe -mminimal-toc -fno-builtin 
-endif
-
-ifdef DBG
-  CFLAGS += -DTIGON3_DEBUG=2
-endif
-
-ifdef SMALL
-  CFLAGS += -Os -DBCM_SMALL_DRV=1
-else
-  CFLAGS += -O2
-endif
-
-all: bcm5700.o
-tar: bcm5700.tgz
-
-bcm5700.o: b57um.o b57proc.o tigon3.o autoneg.o 5701rls.o tcp_seg.o b57diag.o
-	$(LD) -i $^ -o $@
-
-endif # ifeq ($(BCM_KVER),2.6)
-
-bcm5700.4.gz:
-	gzip -c bcm5700.4 > bcm5700.4.gz
-
-ifeq ($(BCM_KVER), 2.6)
-install: default bcm5700.4.gz
-else
-install: $(BCM_DRV) bcm5700.4.gz
-endif
-	@if [ -d $(PREFIX)/lib/modules/$(shell uname -r)/kernel/3rdparty/bcm5700 ];\
-	then rm -f $(PREFIX)/lib/modules/$(shell uname -r)/kernel/3rdparty/bcm5700/$(BCM_DRV).gz;\
-	install -m 444 $(BCM_DRV) $(PREFIX)/lib/modules/$(shell uname -r)/kernel/3rdparty/bcm5700;\
-	elif [ -d $(PREFIX)/lib/modules/$(shell uname -r)/kernel/drivers/addon/bcm5700 ];\
-	then install -m 444 $(BCM_DRV) $(PREFIX)/lib/modules/$(shell uname -r)/kernel/drivers/addon/bcm5700;\
-	elif [ -d $(PREFIX)/lib/modules/$(shell uname -r)/kernel/drivers/net/bcm ];\
-	then install -m 444 $(BCM_DRV) $(PREFIX)/lib/modules/$(shell uname -r)/kernel/drivers/net/bcm;\
-	elif [ -d $(PREFIX)/lib/modules/$(shell uname -r)/kernel ];\
-	then mkdir -p $(PREFIX)/lib/modules/$(shell uname -r)/kernel/drivers/net;\
-	install -m 444 $(BCM_DRV) $(PREFIX)/lib/modules/`uname -r`/kernel/drivers/net;\
-	else mkdir -p $(PREFIX)/lib/modules/$(shell uname -r)/net;\
-	install -m 444 $(BCM_DRV) $(PREFIX)/lib/modules/$(shell uname -r)/net/;\
-	fi
-	@if [ "$(PREFIX)" = "" ]; then /sbin/depmod -a ;\
-	else echo " *** Run '/sbin/depmod -a' to update the module database.";\
-	fi
-	@if [ -d $(PREFIX)/usr/share/man ];\
-	then install -m 444 bcm5700.4.gz $(PREFIX)/usr/share/man/man4;\
-	elif [ -d $(PREFIX)/usr/man ];\
-	then install -m 444 bcm5700.4.gz $(PREFIX)/usr/man/man4;\
-	fi
-
-bcm5700.tgz: Makefile *.[cho] LICENSE RELEASE.TXT
-	tar cfvz $@ $^
-
-rpm: bcm5700.tgz bcm5700.spec
-	cp -p $< ${rpmdir}/SOURCES/
-	rpm -ba bcm5700.spec
-
-.PHONEY: all clean install tar
-
-debug:
-	@make DBG=1
-
-small:
-	@make SMALL=1
-
-clean:
-	-rm -f bcm5700.o bcm5700.ko b57um.o b57proc.o tigon3.o autoneg.o 5701rls.o tcp_seg.o b57diag.o bcm5700.mod.c bcm5700.mod.o
+EXTRA_CFLAGS += -DDBG=0 -DT3_JUMBO_RCV_RCB_ENTRY_COUNT=256 -DNICE_SUPPORT
+EXTRA_CFLAGS += -DPCIX_TARGET_WORKAROUND=1 -DINCLUDE_TBI_SUPPORT -DINCLUDE_5701_AX_FIX=1
+EXTRA_CFLAGS += -I$(P4_TOP_SRCDIR)/include
 
+-include ../Makefile.compat
diff -Naur --exclude .svn bcm5700-8.2.18/mm.h bcm5700-8.2.18-ps/mm.h
--- bcm5700-8.2.18/mm.h	2005-08-02 02:30:49.000000000 +0200
+++ bcm5700-8.2.18-ps/mm.h	2005-10-13 17:02:46.693493076 +0200
@@ -124,9 +124,12 @@
 #define BCM_TASKLET 1
 
 #if HAVE_NETIF_RECEIVE_SKB
+#if 0
+/* ParaStation dont like netif_receive_skb (maybe in the future we do) */
 #define BCM_NAPI_RXPOLL 1
 #undef BCM_TASKLET
 #endif
+#endif
 
 #if defined(CONFIG_PPC64)
 #define BCM_DISCONNECT_AT_CACHELINE 1
diff -Naur --exclude .svn bcm5700-8.2.18/tigon3.h bcm5700-8.2.18-ps/tigon3.h
--- bcm5700-8.2.18/tigon3.h	2005-08-02 02:30:49.000000000 +0200
+++ bcm5700-8.2.18-ps/tigon3.h	2005-10-13 17:09:27.116768928 +0200
@@ -193,14 +193,14 @@
 /* Default coalescing parameters. */
 #ifdef BCM_NAPI_RXPOLL
 #define DEFAULT_RX_COALESCING_TICKS         18
-#define DEFAULT_RX_MAX_COALESCED_FRAMES     6
+#define DEFAULT_RX_MAX_COALESCED_FRAMES     1 /* 6 */
 #else
 #define DEFAULT_RX_COALESCING_TICKS         60
-#define DEFAULT_RX_MAX_COALESCED_FRAMES     15
+#define DEFAULT_RX_MAX_COALESCED_FRAMES     1 /* 15 */
 #endif
 
 #define DEFAULT_TX_COALESCING_TICKS         200
-#define DEFAULT_TX_MAX_COALESCED_FRAMES     35
+#define DEFAULT_TX_MAX_COALESCED_FRAMES     1 /* 35 */
 
 #define MAX_RX_COALESCING_TICKS             500
 #define MAX_TX_COALESCING_TICKS             500
